### DISCUSSION TITLE
Directory search

### DISCUSSION THREAD

**Discussion by TonCherAmi** [MusicPlayerDaemon/MPD#1309] [Q&A]
Hi. Is it possible to search the database for directories? I’ve looked at the [protocol documentation](https://mpd.readthedocs.io/en/stable/protocol.html) and it seems like the only way to do it is to filter through `listall` output which, I understand, is not recommended.

The reason I want to do this is that I organize my music in directories like so:

```
music
├── genre1
│   └── Artist1
│       └── (1970) Album1
│           ├── 01 - Song1
│           └── 02 - Song2
└── genre2
    └── Artist2
        └── (1970) Album2
            ├── 01 - Song1
            └── 02 - Song2
```

I also usually listen to whole albums at once. Therefore I’d like to be able to search, for instance, for the directory called `Album2`  and queue it without having to either navigate the directory tree or specify the full path to the directory by hand.

**Comment by MaxKellermann** [comment]
Not possible. Search results must be songs.
Though you can use a search and match on the `file` property, which includes the directory. https://mpd.readthedocs.io/en/latest/protocol.html#filters

**Reply by TonCherAmi** [reply]
I see, thanks. Would you say this functionality is something worth adding? I'd like to try and implement it, I'm not really sure how it should be done though, for example would this be an entirely new command or an extension of the filter syntax?

**Reply by MaxKellermann** [reply]
Which functionality do you mean? What you ask is already possible, as I already said, e.g.:
```
searchadd "(file =~ '/(1970) Album1/')"
```
This is a single command which adds all songs within a directory called "(1970) Album1". That's exactly what you want, isn't it?

**Reply by TonCherAmi** [reply]
Not quite. Within my client I want to provide a search UI that works like this:
```
Search: alb
---
Results:
(1970) Album1
(1970) Album2
```

```
Search: son
---
Results:
01 - Song1
02 - Song2
```
The user should be able to interact with a search result entry - queue it, descend into it if it is a directory, etc.

The `searchadd` approach wouldn't really allow any interaction like that. It also wouldn't really work for substring searches like in the example above. I guess what I want is a basename search that works for both regular files and directories. This would be trivial to implement by parsing `listall` output but that's less than ideal of course.

**Reply by MaxKellermann** [reply]
Why do you mention `listall` all the time, when you can just let `search` do it?
```
search "(file =~ 'alb')"
```
I don't quite understand what you want, what is not already possible.

**Reply by MaxKellermann** [reply]
And to reduce network load (because you're not interested in tag values), you can ask MPD to omit all tags by sending `tagtypes clear` (see https://mpd.readthedocs.io/en/latest/protocol.html#connection-settings) before sending `search ...`.

**Reply by TonCherAmi** [reply]
>  `search "(file =~ 'alb')"`

This would just return all of the songs inside those albums and there could be *a lot* of them. The user would then have to queue all of them manually instead of selecting a single entry. There could also be an overlap between a search string and a name of a genre which could result in thousands of unwanted results, whereas if we only searched for the basename there would be one additional directory present.

**Reply by MaxKellermann** [reply]
I still don't understand you.

> This would just return all of the songs inside those albums and there could be _a lot_ of them.

Your other idea to use `listall` returns **all** songs, which is **a lot lot more** than just those with matching containing directory names. My suggestion is much better than that, isn't it?

> The user would then have to queue all of them manually instead of selecting a single entry.

This I don't grasp at all.
Your client can decide to show just the parent directory names in the UI, and not individual songs.
What would your `listall` approach do?

> There could also be an overlap between a search string and a name of a genre which could result in thousands of unwanted results, whereas if we only searched for the basename there would be one additional directory present.

That was just an example. Improving that regexp so it doesn't match on genres is left as exercise for you. I just gave you the basic concept, and I trusted that you'd refine it. But hey .... here you are:
```
search "(file =~ '/.*alb')"
```
This regexp would search all path components but the first one.

**Reply by TonCherAmi** [reply]
> Your client can decide to show just the parent directory names in the UI, and not individual songs.

You're suggesting that the client should examine every path segment in every returned path for substring matches?
So if I do a `search "(file =~ 'alb')"` I get
```
genre1/Artist1/(1970) Album1/01 - Song1
genre1/Artist1/(1970) Album1/02 - Song2
genre2/Artist2/(1970) Album2/01 - Song1
genre2/Artist2/(1970) Album2/02 - Song2
```
and then I do something like this?
```
for result in search(query)
  for segment in result.path.split('/')
    path_to_return.append("/$segment")
    
    if segment.includes(search_string)
      break
  
  results.push(path_to_return)
  ...
```

> What would your `listall` approach do?

I mean that would be pretty straightforward:
```
for entry in listall()
  if basename(entry.path).includes(search_string)
    results.push(entry.path)
```

**Reply by MaxKellermann** [reply]
> You're suggesting that the client should examine every path segment in every returned path for substring matches?

No, no, no! Why would you examine for substring matches, when MPD has already done that?

> I mean that would be pretty straightforward:
> 
> ```
> for entry in listall()
>   if basename(entry.path).includes(search_string)
>     results.push(entry.path)
> ```

But ... that shows all songs, not directories, which is what you previously described that you explicitly do **not** want! You're comparing apples to oranges.

The equivalent code sample with `search` would be:
```
for entry in search(search_string)
  results.push(entrypath)
```
Understand? Do you see that this is both simpler *and* more efficient?

**Reply by TonCherAmi** [reply]
> But ... that shows all songs, not directories, which is what you previously described that you explicitly do not want! You're comparing apples to oranges.

Wait, huh? Here I just ran `listall` and the directories are there?

![image](https://user-images.githubusercontent.com/13941584/138775951-62f9f9c2-adde-4af5-bd28-c75ce62f81ac.png)

**Reply by MaxKellermann** [reply]
And all songs are there as well.

**Reply by TonCherAmi** [reply]
Yes, that works perfectly for me, I want directories *and songs* whose basename matches the search string.

**Reply by MaxKellermann** [reply]
`search` gives you the songs, and the directory names are trivial to deduce from there, aren't they?

**Reply by TonCherAmi** [reply]
So I need to do this after all?

> You're suggesting that the client should examine every path segment in every returned path for substring matches?

I mean it's not pretty but I guess it beats caching the entire database on the client..

**Reply by MaxKellermann** [reply]
> So I need to do this after all?
> 
> > You're suggesting that the client should examine every path segment in every returned path for substring matches?

You asked this question already, and remember my reply? Here's a copy of it:
> No no no!

... which effectively means "No, you do not need to do this after all", just in case three times "no" wasn't clear enough. For further explanation, scroll up to re-read my full text response.

**Reply by TonCherAmi** [reply]
Well, I see no other way of getting the right directories from `search` :man_shrugging: 
Simply showing parent directories won't work because there could be song matches as well.
`search` just doesn't do what I want and implementing it client-side seems finicky, that's why I asked whether this could be added to MPD itself.
